jsonapi/marshal.go

```diff
@@ -113,107 +118,108 @@ func MarshalToStruct(data interface{}, information ServerInformation) (Document,
-	result := Document{}
+func marshalSlice(data interface{}, information ServerInformation) (*Document, error) {
+	result := &Document{}
 
 	val := reflect.ValueOf(data)
-	dataElements := []Data{}
+	dataElements := make([]Data, val.Len())
 	var referencedStructs []MarshalIdentifier
 
 	for i := 0; i < val.Len(); i++ {
 		k := val.Index(i).Interface()
 		element, ok := k.(MarshalIdentifier)
 		if !ok {
-			return result, errors.New("all elements within the slice must implement api2go.MarshalIdentifier")
+			return nil, errors.New("all elements within the slice must implement api2go.MarshalIdentifier")
 		}
 
-		content, err := marshalData(element, information)
+		err := marshalData(element, &dataElements[i], information)
 		if err != nil {
-			return result, err
+			return nil, err
 		}
 
-		dataElements = append(dataElements, *content)
-
 		included, ok := k.(MarshalIncludedRelations)
 		if ok {
 			referencedStructs = append(referencedStructs, included.GetReferencedStructs()...)
 		}
 	}
 
-	includedElements := reduceDuplicates(referencedStructs, information, marshalData)
+	includedElements, err := filterDuplicates(referencedStructs, information)
+	if err != nil {
+		return nil, err
+	}
 
-	//data key is always present
 	result.Data = &DataContainer{
 		DataArray: dataElements,
 	}
-	if includedElements != nil && len(*includedElements) > 0 {
-		result.Included = *includedElements
+
+	if includedElements != nil && len(includedElements) > 0 {
+		result.Included = includedElements
 	}
 
 	return result, nil
 }
 
-// reduceDuplicates eliminates duplicate MarshalIdentifier from input and calls `method` on every unique MarshalIdentifier
-func reduceDuplicates(
-	input []MarshalIdentifier,
-	information ServerInformation,
-	method func(MarshalIdentifier, ServerInformation) (*Data, error),
-) *[]Data {
+func filterDuplicates(input []MarshalIdentifier, information ServerInformation) ([]Data, error) {
 	alreadyIncluded := map[string]map[string]bool{}
 	includedElements := []Data{}
 
 	for _, referencedStruct := range input {
 		structType := getStructType(referencedStruct)
+
 		if alreadyIncluded[structType] == nil {
 			alreadyIncluded[structType] = make(map[string]bool)
 		}
 
 		if !alreadyIncluded[structType][referencedStruct.GetID()] {
-			marshalled, _ := method(referencedStruct, information)
-			includedElements = append(includedElements, *marshalled)
+			var data Data
+			err := marshalData(referencedStruct, &data, information)
+			if err != nil {
+				return nil, err
+			}
+
+			includedElements = append(includedElements, data)
 			alreadyIncluded[structType][referencedStruct.GetID()] = true
 		}
 	}
 
-	return &includedElements
+	return includedElements, nil
 }
 
-func marshalData(element MarshalIdentifier, information ServerInformation) (*Data, error) {
-	var err error
-	result := &Data{}
-
+func marshalData(element MarshalIdentifier, data *Data, information ServerInformation) error {
```

Comment:
this is better because we can already write into a pre allocated `Data` struct instead of creating and returning a new one, right?